### ISSUE TITLE
Converting UK biobank SWI protocol to BIDS format

### ISSUE THREAD

**Issue by karllandheer** [QSMxT/QSMxT#115]
Hello, thank you very much for this great package. I am interested in running it on UK biobank SWI data to generate QSM maps and I have a few questions I was hoping you could help me with.

1) Is there an easy way to convert UK biobank data into the required BIDS format for QSMxT? I tried using dicom-sort and dicom-convert, however the resulting nifti images were certainly wrong. My guess is because the individual coil data are provided by default? It's possible I did something wrong as well. A sample UK biobank SWI image dataset can be obtained here: https://biobank.ndph.ox.ac.uk/showcase/refer.cgi?id=32579

2) Is there a recommended "general purpose" algorithm that you recommend? I am very new to QSM, and as I can see there's a bunch of algorithms included in QSMxT (which is great!) which all have their own benefits and drawbacks, but in general which one would you recommend for UK biobank data?

3) How long does this recommend algorithm take to run? As you can imagine compute can become an issue when we're dealing with so many datasets. 

Any help would be greatly appreciated!

**Comment by karllandheer** [comment]
Hello, I've been muddling my way through 1). I think the easiest way is to convert the dicoms to nifti and then grab the corresponding sum-of-squares images that are available. I then ran nifti-convert on the folder which has the two sum of squares images (both magnitude and phase). The output is a screenshot of dataset_qsmxt.csv, which I believe I have to input and then I should be able to run qsmxt on my data. I have a few questions about these input

DerivativePipeline: is this the algorithm I want to run?
Sub: I assume that's the subject number, obviously 1 in my case?
Ses: I assume thats session number
Acq: Not sure
ce: not sure
rec: not sure
run: not sure
mod: not sure
echo: is this echo number? For me this would either be 1 or 2? Or is it total number of echoes (2 for both)?
flip: Excitation flip angle in degrees?
inv: not sure
mt: not sure
part: is this magnitude or phase? It seemed to grab phase for the _ph images but not magnitude for the others. 
desc: description? 
MagneticFieldStrength: Field strength in Tesla I assume?
EchoTime: Echo time in ms I assume?

Do you have a sample dataset_qsmxt.csv floating around? Sorry if this is answered somewhere else, but I'm a bit new to BIDS and it seems BIDS is less well described for SWI than T1. 

<img width="1360" alt="Screenshot 2024-03-22 at 4 48 06 PM" src="https://github.com/QSMxT/QSMxT/assets/76697225/7053c248-74a5-410b-a401-44294e6907f7">

**Comment by astewartau** [comment]
Hi @karllandheer, and thanks for your interest in using QSMxT! 

# 1. Conversion to BIDS
You can convert UK BioBank into the BIDS format, and you are on the right track. I have found the best way to do this is by starting with the NIfTI formatted data from UK BioBank, and using `nifti-convert` to convert to BIDS, filling out the spreadsheet as you have started. Unfortunately, you need some expertise with the BIDS specification to know how to do this correctly - other tools can assist with BIDS conversions and are more user-friendly, but I don't know of any that reliably work with QSM data. I have documented a correct example here.

I downloaded subject 1000219 data types 20251, 20252, and 26301:

```bash
~/neurodesktop-storage/data/2023-BioBank-QSM-Data: tree 1000219/
1000219/
├── 1000219_20251_2_0
│   └── SWI
│       ├── brain_mask.nii.gz
│       ├── filtered_phase.nii.gz
│       ├── SOS_TE1.nii.gz
│       ├── SOS_TE2.nii.gz
│       ├── SWI.nii.gz
│       ├── SWI_to_T1.mat
│       ├── SWI_TOTAL_MAG.json
│       ├── SWI_TOTAL_MAG.nii.gz
│       ├── SWI_TOTAL_MAG_orig.nii.gz
│       ├── SWI_TOTAL_MAG_TE2.json
│       ├── SWI_TOTAL_MAG_TE2_orig.nii.gz
│       ├── SWI_TOTAL_MAG_to_T1.nii.gz
│       ├── SWI_TOTAL_PHA.json
│       ├── SWI_TOTAL_PHA_TE2.json
│       ├── T1_to_SWI.mat
│       ├── T2star.nii.gz
│       └── T2star_to_T1.nii.gz
├── 1000219_20252_2_0
│   └── T1
│       ├── T1_brain_mask.nii.gz
│       ├── T1_brain.nii.gz
│       ├── T1_brain_to_MNI.nii.gz
│       ├── T1_fast
│       │   ├── T1_brain_bias.nii.gz
│       │   ├── T1_brain_pve_0.nii.gz
│       │   ├── T1_brain_pve_1.nii.gz
│       │   ├── T1_brain_pve_2.nii.gz
│       │   └── T1_brain_seg.nii.gz
│       ├── T1_first
│       │   ├── T1_first_all_fast_firstseg.nii.gz
│       │   ├── T1_first-BrStem_first.bvars
│       │   ├── T1_first-BrStem_first.vtk
│       │   ├── T1_first-L_Accu_first.bvars
│       │   ├── T1_first-L_Accu_first.vtk
│       │   ├── T1_first-L_Amyg_first.bvars
│       │   ├── T1_first-L_Amyg_first.vtk
│       │   ├── T1_first-L_Caud_first.bvars
│       │   ├── T1_first-L_Caud_first.vtk
│       │   ├── T1_first-L_Hipp_first.bvars
│       │   ├── T1_first-L_Hipp_first.vtk
│       │   ├── T1_first-L_Pall_first.bvars
│       │   ├── T1_first-L_Pall_first.vtk
│       │   ├── T1_first-L_Puta_first.bvars
│       │   ├── T1_first-L_Puta_first.vtk
│       │   ├── T1_first-L_Thal_first.bvars
│       │   ├── T1_first-L_Thal_first.vtk
│       │   ├── T1_first-R_Accu_first.bvars
│       │   ├── T1_first-R_Accu_first.vtk
│       │   ├── T1_first-R_Amyg_first.bvars
│       │   ├── T1_first-R_Amyg_first.vtk
│       │   ├── T1_first-R_Caud_first.bvars
│       │   ├── T1_first-R_Caud_first.vtk
│       │   ├── T1_first-R_Hipp_first.bvars
│       │   ├── T1_first-R_Hipp_first.vtk
│       │   ├── T1_first-R_Pall_first.bvars
│       │   ├── T1_first-R_Pall_first.vtk
│       │   ├── T1_first-R_Puta_first.bvars
│       │   ├── T1_first-R_Puta_first.vtk
│       │   ├── T1_first-R_Thal_first.bvars
│       │   └── T1_first-R_Thal_first.vtk
│       ├── T1.json
│       ├── T1.nii.gz
│       ├── T1_orig_defaced.nii.gz
│       ├── T1_unbiased_brain.nii.gz
│       └── transforms
│           ├── T1_to_MNI_linear.mat
│           └── T1_to_MNI_warp_coef.nii.gz
└── 1000219_26301_2_0
    └── SWI
        └── QSM
            ├── PHASE_TE1.nii.gz
            ├── PHASE_TE2.nii.gz
            ├── QSM_CSFref.nii.gz
            ├── QSM_CSF.txt
            ├── QSM_mask.nii.gz
            ├── QSM.nii.gz
            └── QSM_ud.nii.gz

11 directories, 69 files
```

Then, I started the conversion to bids using `nifti-convert`:

```bash
~/neurodesktop-storage/data/2023-BioBank-QSM-Data: nifti-convert 1000219 bids
[INFO]: Running QSMxT 6.4.2 (linked installation)
[INFO]: Command: /home/ashley/neurodesktop-storage/miniconda3/envs/qsmxt/bin/nifti-convert 1000219 bids
[INFO]: Python interpreter: /home/ashley/neurodesktop-storage/miniconda3/envs/qsmxt/bin/python
[INFO]: Finding NIfTI files...
[INFO]: 30 NIfTI files found.
[INFO]: Extracting details from filenames...
[INFO]: Done reading details.
[INFO]: Updating details with JSON header information...
[INFO]: Done reading JSON header files.
[INFO]: RUN AGAIN AFTER ENTERING RELEVANT BIDS INFORMATION TO /home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/dataset_qsmxt.csv.
[INFO]: Finished
```

This generated a spreadsheet:

![image](https://github.com/QSMxT/QSMxT/assets/9957135/49d35f97-b0aa-42f3-bb51-e3160dd6d9b1)

Which I filled out as follows: 

![image](https://github.com/QSMxT/QSMxT/assets/9957135/02d272f0-7e8c-4093-88a3-39e07dbfe96e)

Running again completes the conversion:

```bash
(qsmxt) ~/neurodesktop-storage/data/2023-BioBank-QSM-Data: nifti-convert 1000219/ bids
[INFO]: Running QSMxT 6.4.2 (linked installation)
[INFO]: Command: /home/ashley/neurodesktop-storage/miniconda3/envs/qsmxt/bin/nifti-convert 1000219/ bids
[INFO]: Python interpreter: /home/ashley/neurodesktop-storage/miniconda3/envs/qsmxt/bin/python
[INFO]: CSV spreadsheet '/home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/dataset_qsmxt.csv' found! Reading...
[INFO]: CSV spreadsheet loaded.
[INFO]: Computing new NIfTI file names and locations...
[INFO]: New NIfTI file names and locations determined.
Summary of identified files and proposed new names (following BIDS standard):
SWI_TOTAL_MAG_TE2_orig.nii.gz 
	 -> sub-1000219_ses-20190507_acq-swi_echo-2_part-mag_MEGRE.nii.gz
SWI_TOTAL_MAG_orig.nii.gz 
	 -> sub-1000219_ses-20190507_acq-swi_echo-1_part-mag_MEGRE.nii.gz
T1_orig_defaced.nii.gz 
	 -> sub-1000219_ses-20190507_acq-t1w_T1w.nii.gz
PHASE_TE1.nii.gz 
	 -> sub-1000219_ses-20190507_acq-swi_echo-1_part-phase_MEGRE.nii.gz
PHASE_TE2.nii.gz 
	 -> sub-1000219_ses-20190507_acq-swi_echo-2_part-phase_MEGRE.nii.gz
QSM_mask.nii.gz 
	 -> sub-1000219_ses-20190507_acq-swi_mask.nii.gz
Confirm copy + renames? (n for no): 

[INFO]: Copying NIfTI files to new locations with new names...
[INFO]: Done copying NIfTI files.
[INFO]: Copying JSON header files if present and generating them if needed...
[INFO]: Automatically generated JSON header file '/home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/sub-1000219/ses-20190507/anat/sub-1000219_ses-20190507_acq-swi_echo-2_part-mag_MEGRE.json'
[INFO]: Automatically generated JSON header file '/home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/sub-1000219/ses-20190507/anat/sub-1000219_ses-20190507_acq-swi_echo-1_part-mag_MEGRE.json'
[INFO]: Automatically generated JSON header file '/home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/sub-1000219/ses-20190507/anat/sub-1000219_ses-20190507_acq-t1w_T1w.json'
[INFO]: Automatically generated JSON header file '/home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/sub-1000219/ses-20190507/anat/sub-1000219_ses-20190507_acq-swi_echo-1_part-phase_MEGRE.json'
[INFO]: Automatically generated JSON header file '/home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/sub-1000219/ses-20190507/anat/sub-1000219_ses-20190507_acq-swi_echo-2_part-phase_MEGRE.json'
[INFO]: Automatically generated JSON header file '/home/ashley/neurodesktop-storage/data/2023-BioBank-QSM-Data/bids/derivatives/biobank/sub-1000219/ses-20190507/anat/sub-1000219_ses-20190507_acq-swi_mask.json'
[INFO]: Done copying and generating JSON header files.
[INFO]: Generating details for BIDS datset_description.json...
[INFO]: Writing BIDS dataset_description.json...
[INFO]: Writing BIDS .bidsignore file...
[INFO]: Writing BIDS dataset README...
[INFO]: Finished
```

With the final folder appearing as follows:

```bash
(qsmxt) ~/neurodesktop-storage/data/2023-BioBank-QSM-Data: tree bids
bids
├── dataset_description.json
├── dataset_qsmxt.csv
├── derivatives
│   └── biobank
│       └── sub-1000219
│           └── ses-20190507
│               └── anat
│                   ├── sub-1000219_ses-20190507_acq-swi_mask.json
│                   └── sub-1000219_ses-20190507_acq-swi_mask.nii.gz
├── log_2024-03-25_16-53-20248899.txt
├── log_2024-03-25_16-53-33292075.txt
├── README
├── references.txt
└── sub-1000219
    └── ses-20190507
        └── anat
            ├── sub-1000219_ses-20190507_acq-swi_echo-1_part-mag_MEGRE.json
            ├── sub-1000219_ses-20190507_acq-swi_echo-1_part-mag_MEGRE.nii.gz
            ├── sub-1000219_ses-20190507_acq-swi_echo-1_part-phase_MEGRE.json
            ├── sub-1000219_ses-20190507_acq-swi_echo-1_part-phase_MEGRE.nii.gz
            ├── sub-1000219_ses-20190507_acq-swi_echo-2_part-mag_MEGRE.json
            ├── sub-1000219_ses-20190507_acq-swi_echo-2_part-mag_MEGRE.nii.gz
            ├── sub-1000219_ses-20190507_acq-swi_echo-2_part-phase_MEGRE.json
            ├── sub-1000219_ses-20190507_acq-swi_echo-2_part-phase_MEGRE.nii.gz
            ├── sub-1000219_ses-20190507_acq-t1w_T1w.json
            └── sub-1000219_ses-20190507_acq-t1w_T1w.nii.gz

9 directories, 18 files
```

# 2. Recommended algorithm

QSMxT has several pipelines you can run, each applying different collections of algorithms and parameter values. I would usually recommend using the default pipeline (GRE) to start with and branching out depending on the data - e.g. the EPI pipeline tends to deal with the lower SNR of accelerated acquisitions better than the GRE pipeline; the body pipeline was optimised for a prostate imaging application and doesn't assume brain anatomy etc.

After some limited testing using the BioBank data, I found that the result improved when we added a BET mask to the default threshold-based masking algorithm, which you can do by applying the `--add_bet` flag (or enabling the equivalent setting using the interactive menus). I used the following command:

```bash
qsmxt bids qsm --add_bet --auto_yes
```

![image](https://github.com/QSMxT/QSMxT/assets/9957135/efcccd65-6a60-45e6-a062-327d58af97fa)

# 3. Runtime

For one subject, I just timed this at about 11 minutes for the one subject. However, if you have access to a high-performance cluster (HPC) with SLURM or PBS, QSMxT can submit each algorithmic step as a job, parallelising all of the processing and making this much more feasible to run in a short amount of time. To do this, follow the HPC install instructions [here](https://qsmxt.github.io/QSMxT/installation#hpc-installation), then when you run, use the --slurm or --pbs flag. For SLURM, use:

```bash
qsmxt bids qsm --add_bet --auto_yes --slurm ACCOUNT_STRING SLURM_QUEUE
```

Replacing `ACCOUNT_STRING` with the account string or group with SLURM permissions enabled and `SLURM_QUEUE` with the queue name to which jobs should be submitted. For PBS, it is similar:

```bash
qsmxt bids qsm --add_bet --auto_yes --pbs ACCOUNT_STRING
```

Hope that helps!

**Comment by karllandheer** [comment]
Hello, yes that was indeed the issue. I am having another issue now. Working on figuring out what the issue is... Here's the error log:


(UKB) karl.landheer@C02F203JMD6R UKB_Brain_Pipeline % qsmxt /Users/karl.landheer/Downloads/bids_output/ /Users/karl.landheer/Downloads/qsm_output --auto_yes                
[INFO]: Loading previous QSMxT settings...
[INFO]: QSMxT v6.4.2
[INFO]: Python interpreter: /Users/karl.landheer/opt/anaconda3/envs/ukb/bin/python
[INFO]: Command: qsmxt /Users/karl.landheer/Downloads/bids_output /Users/karl.landheer/Downloads/qsm_output --auto_yes
[INFO]: Creating QSMxT workflow for sub-1.ses-1.acq-swi...
[INFO]: Running using MultiProc plugin with n_procs=12
240325-10:12:29,589 nipype.workflow INFO:
	 Workflow workflow settings: ['check', 'execution', 'logging', 'monitoring']
240325-10:12:29,644 nipype.workflow INFO:
	 Running in parallel.
240325-10:12:29,648 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 4 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:31,652 nipype.workflow INFO:
	 [MultiProc] Running 3 tasks, and 2 jobs ready. Free memory (GB): 13.80/14.40, Free processors: 9/12.
                     Currently running:
                       * workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_as-canonical
                       * workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_read-nii
                       * workflow.sub-1.ses-1.qsmxt_acq-swi.func_read-json-se
240325-10:12:33,654 nipype.workflow INFO:
	 [MultiProc] Running 5 tasks, and 0 jobs ready. Free memory (GB): 13.40/14.40, Free processors: 7/12.
                     Currently running:
                       * _func_read-json-me1
                       * _func_read-json-me0
                       * workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_as-canonical
                       * workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_read-nii
                       * workflow.sub-1.ses-1.qsmxt_acq-swi.func_read-json-se
240325-10:12:41,404 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_read-nii" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_read-nii".
240325-10:12:41,404 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.func_read-json-se" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/func_read-json-se".
240325-10:12:41,404 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_as-canonical" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_as-canonical".
240325-10:12:41,404 nipype.workflow INFO:
	 [Node] Setting-up "_func_read-json-me0" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/func_read-json-me/mapflow/_func_read-json-me0".
240325-10:12:41,404 nipype.workflow INFO:
	 [Node] Setting-up "_func_read-json-me1" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/func_read-json-me/mapflow/_func_read-json-me1".
240325-10:12:41,410 nipype.workflow INFO:
	 [Node] Executing "func_read-json-se" <nipype.interfaces.utility.wrappers.Function>
240325-10:12:41,411 nipype.workflow INFO:
	 [Node] Executing "nibabel_read-nii" <nipype.interfaces.utility.wrappers.Function>
240325-10:12:41,411 nipype.workflow INFO:
	 [Node] Executing "nibabel_as-canonical" <nipype.interfaces.utility.wrappers.Function>
240325-10:12:41,411 nipype.workflow INFO:
	 [Node] Executing "_func_read-json-me0" <nipype.interfaces.utility.wrappers.Function>
240325-10:12:41,411 nipype.workflow INFO:
	 [Node] Executing "_func_read-json-me1" <nipype.interfaces.utility.wrappers.Function>
240325-10:12:41,457 nipype.workflow INFO:
	 [Node] Finished "_func_read-json-me1", elapsed time 0.003029s.
240325-10:12:41,457 nipype.workflow INFO:
	 [Node] Finished "func_read-json-se", elapsed time 0.002512s.
240325-10:12:41,457 nipype.workflow INFO:
	 [Node] Finished "_func_read-json-me0", elapsed time 0.00203s.
240325-10:12:41,461 nipype.workflow INFO:
	 [Node] Finished "nibabel_read-nii", elapsed time 0.006699s.
240325-10:12:41,480 nipype.workflow INFO:
	 [Node] Finished "nibabel_as-canonical", elapsed time 0.025367s.
240325-10:12:41,663 nipype.workflow INFO:
	 [Job 1] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.func_read-json-se).
240325-10:12:41,675 nipype.workflow INFO:
	 [Job 2] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_read-nii).
240325-10:12:41,676 nipype.workflow INFO:
	 [Job 3] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_as-canonical).
240325-10:12:41,678 nipype.workflow INFO:
	 [Job 28] Completed (_func_read-json-me0).
240325-10:12:41,680 nipype.workflow INFO:
	 [Job 29] Completed (_func_read-json-me1).
240325-10:12:41,682 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 2 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:41,744 nipype.workflow INFO:
	 [Node] Setting-up "_func_read-json-me0" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/func_read-json-me/mapflow/_func_read-json-me0".
240325-10:12:41,751 nipype.workflow INFO:
	 [Node] Cached "_func_read-json-me0" - collecting precomputed outputs
240325-10:12:41,751 nipype.workflow INFO:
	 [Node] "_func_read-json-me0" found cached.
240325-10:12:41,752 nipype.workflow INFO:
	 [Node] Setting-up "_func_read-json-me1" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/func_read-json-me/mapflow/_func_read-json-me1".
240325-10:12:41,758 nipype.workflow INFO:
	 [Node] Cached "_func_read-json-me1" - collecting precomputed outputs
240325-10:12:41,759 nipype.workflow INFO:
	 [Node] "_func_read-json-me1" found cached.
240325-10:12:43,667 nipype.workflow INFO:
	 [Job 0] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.func_read-json-me).
240325-10:12:43,678 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 2 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:43,737 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_scale-phase0" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_scale-phase/mapflow/_nibabel_numpy_scale-phase0".
240325-10:12:43,738 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_scale-phase1" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_scale-phase/mapflow/_nibabel_numpy_scale-phase1".
240325-10:12:43,742 nipype.workflow INFO:
	 [Node] Executing "_nibabel_numpy_scale-phase0" <qsmxt.interfaces.nipype_interface_processphase.ScalePhaseInterface>
240325-10:12:43,743 nipype.workflow INFO:
	 [Node] Executing "_nibabel_numpy_scale-phase1" <qsmxt.interfaces.nipype_interface_processphase.ScalePhaseInterface>
240325-10:12:44,409 nipype.workflow INFO:
	 [Node] Finished "_nibabel_numpy_scale-phase0", elapsed time 0.650205s.
240325-10:12:44,420 nipype.workflow INFO:
	 [Node] Finished "_nibabel_numpy_scale-phase1", elapsed time 0.662092s.
240325-10:12:45,670 nipype.workflow INFO:
	 [Job 30] Completed (_nibabel_numpy_scale-phase0).
240325-10:12:45,677 nipype.workflow INFO:
	 [Job 31] Completed (_nibabel_numpy_scale-phase1).
240325-10:12:45,679 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 1 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:45,743 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_scale-phase0" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_scale-phase/mapflow/_nibabel_numpy_scale-phase0".
240325-10:12:45,748 nipype.workflow INFO:
	 [Node] Cached "_nibabel_numpy_scale-phase0" - collecting precomputed outputs
240325-10:12:45,749 nipype.workflow INFO:
	 [Node] "_nibabel_numpy_scale-phase0" found cached.
240325-10:12:45,750 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_scale-phase1" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_scale-phase/mapflow/_nibabel_numpy_scale-phase1".
240325-10:12:45,755 nipype.workflow INFO:
	 [Node] Cached "_nibabel_numpy_scale-phase1" - collecting precomputed outputs
240325-10:12:45,755 nipype.workflow INFO:
	 [Node] "_nibabel_numpy_scale-phase1" found cached.
240325-10:12:47,672 nipype.workflow INFO:
	 [Job 4] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_numpy_scale-phase).
240325-10:12:47,678 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 1 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:49,677 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 2 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:49,731 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_nilearn_axial-resampling0" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_nilearn_axial-resampling/mapflow/_nibabel_numpy_nilearn_axial-resampling0".
240325-10:12:49,732 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_nilearn_axial-resampling1" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_nilearn_axial-resampling/mapflow/_nibabel_numpy_nilearn_axial-resampling1".
240325-10:12:49,737 nipype.workflow INFO:
	 [Node] Executing "_nibabel_numpy_nilearn_axial-resampling0" <qsmxt.interfaces.nipype_interface_axialsampling.AxialSamplingInterface>
240325-10:12:49,737 nipype.workflow INFO:
	 [Node] Executing "_nibabel_numpy_nilearn_axial-resampling1" <qsmxt.interfaces.nipype_interface_axialsampling.AxialSamplingInterface>
240325-10:12:51,679 nipype.workflow INFO:
	 [MultiProc] Running 2 tasks, and 0 jobs ready. Free memory (GB): 6.40/14.40, Free processors: 10/12.
                     Currently running:
                       * _nibabel_numpy_nilearn_axial-resampling1
                       * _nibabel_numpy_nilearn_axial-resampling0
240325-10:12:55,540 nipype.workflow INFO:
	 [Node] Finished "_nibabel_numpy_nilearn_axial-resampling1", elapsed time 5.782038s.
240325-10:12:55,558 nipype.workflow INFO:
	 [Node] Finished "_nibabel_numpy_nilearn_axial-resampling0", elapsed time 5.799978s.
240325-10:12:55,686 nipype.workflow INFO:
	 [Job 32] Completed (_nibabel_numpy_nilearn_axial-resampling0).
240325-10:12:55,687 nipype.workflow INFO:
	 [Job 33] Completed (_nibabel_numpy_nilearn_axial-resampling1).
240325-10:12:55,689 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 1 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:55,753 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_nilearn_axial-resampling0" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_nilearn_axial-resampling/mapflow/_nibabel_numpy_nilearn_axial-resampling0".
240325-10:12:55,759 nipype.workflow INFO:
	 [Node] Cached "_nibabel_numpy_nilearn_axial-resampling0" - collecting precomputed outputs
240325-10:12:55,759 nipype.workflow INFO:
	 [Node] "_nibabel_numpy_nilearn_axial-resampling0" found cached.
240325-10:12:55,760 nipype.workflow INFO:
	 [Node] Setting-up "_nibabel_numpy_nilearn_axial-resampling1" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_nilearn_axial-resampling/mapflow/_nibabel_numpy_nilearn_axial-resampling1".
240325-10:12:55,782 nipype.workflow INFO:
	 [Node] Cached "_nibabel_numpy_nilearn_axial-resampling1" - collecting precomputed outputs
240325-10:12:55,782 nipype.workflow INFO:
	 [Node] "_nibabel_numpy_nilearn_axial-resampling1" found cached.
240325-10:12:57,691 nipype.workflow INFO:
	 [Job 5] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.nibabel_numpy_nilearn_axial-resampling).
240325-10:12:57,693 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 5 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:12:57,751 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.mask_workflow.nibabal-numpy_combine-magnitude" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mask_workflow/nibabal-numpy_combine-magnitude".
240325-10:12:57,751 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.mrt_romeo_combine-phase" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mrt_romeo_combine-phase".
240325-10:12:57,751 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.mask_workflow.func_get-first-phase" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mask_workflow/func_get-first-phase".
240325-10:12:57,751 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.mask-intermediate_workflow.nibabal-numpy_combine-magnitude" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mask-intermediate_workflow/nibabal-numpy_combine-magnitude".
240325-10:12:57,751 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.mask-intermediate_workflow.func_get-first-phase" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mask-intermediate_workflow/func_get-first-phase".
240325-10:12:57,783 nipype.workflow INFO:
	 [Node] Executing "nibabal-numpy_combine-magnitude" <qsmxt.interfaces.nipype_interface_combinemagnitude.CombineMagnitudeInterface>
240325-10:12:57,783 nipype.workflow INFO:
	 [Node] Executing "nibabal-numpy_combine-magnitude" <qsmxt.interfaces.nipype_interface_combinemagnitude.CombineMagnitudeInterface>
240325-10:12:57,783 nipype.workflow INFO:
	 [Node] Executing "func_get-first-phase" <nipype.interfaces.utility.wrappers.Function>
240325-10:12:57,784 nipype.workflow INFO:
	 [Node] Executing "func_get-first-phase" <nipype.interfaces.utility.wrappers.Function>
240325-10:12:57,786 nipype.workflow INFO:
	 [Node] Finished "func_get-first-phase", elapsed time 0.000747s.
240325-10:12:57,787 nipype.workflow INFO:
	 [Node] Finished "func_get-first-phase", elapsed time 0.000754s.
240325-10:12:57,787 nipype.workflow INFO:
	 [Node] Executing "mrt_romeo_combine-phase" <qsmxt.interfaces.nipype_interface_romeo.RomeoB0Interface>
240325-10:12:59,451 nipype.workflow INFO:
	 [Node] Finished "nibabal-numpy_combine-magnitude", elapsed time 1.66557s.
240325-10:12:59,472 nipype.workflow INFO:
	 [Node] Finished "nibabal-numpy_combine-magnitude", elapsed time 1.6877689999999999s.
240325-10:12:59,692 nipype.workflow INFO:
	 [Job 7] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.mask_workflow.nibabal-numpy_combine-magnitude).
240325-10:12:59,693 nipype.workflow INFO:
	 [Job 8] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.mask_workflow.func_get-first-phase).
240325-10:12:59,695 nipype.workflow INFO:
	 [Job 9] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.mask-intermediate_workflow.nibabal-numpy_combine-magnitude).
240325-10:12:59,696 nipype.workflow INFO:
	 [Job 10] Completed (workflow.sub-1.ses-1.qsmxt_acq-swi.mask-intermediate_workflow.func_get-first-phase).
240325-10:12:59,698 nipype.workflow INFO:
	 [MultiProc] Running 1 tasks, and 1 jobs ready. Free memory (GB): 9.36/14.40, Free processors: 13/12.
                     Currently running:
                       * workflow.sub-1.ses-1.qsmxt_acq-swi.mrt_romeo_combine-phase
240325-10:12:59,747 nipype.workflow INFO:
	 [Node] Setting-up "workflow.sub-1.ses-1.qsmxt_acq-swi.mask_workflow.romeo-voxelquality" in "/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mask_workflow/romeo-voxelquality".
240325-10:12:59,762 nipype.workflow INFO:
	 [Node] Executing "romeo-voxelquality" <qsmxt.interfaces.nipype_interface_phaseweights.RomeoMaskingInterface>
240325-10:12:59,943 nipype.workflow INFO:
	 [Node] Finished "mrt_romeo_combine-phase", elapsed time 2.144515s.
240325-10:12:59,943 nipype.workflow WARNING:
	 Storing result file without outputs
240325-10:12:59,958 nipype.workflow WARNING:
	 [Node] Error on "workflow.sub-1.ses-1.qsmxt_acq-swi.mrt_romeo_combine-phase" (/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mrt_romeo_combine-phase)
240325-10:12:59,959 nipype.workflow INFO:
	 [Node] Finished "romeo-voxelquality", elapsed time 0.184948s.
240325-10:12:59,960 nipype.workflow WARNING:
	 [Node] Error on "workflow.sub-1.ses-1.qsmxt_acq-swi.mask_workflow.romeo-voxelquality" (/Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mask_workflow/romeo-voxelquality)
240325-10:13:01,698 nipype.workflow ERROR:
	 Node mrt_romeo_combine-phase failed to run on host C02F203JMD6R.
240325-10:13:01,700 nipype.workflow ERROR:
	 Saving crash info to /Users/karl.landheer/PythonProjects/RGCi_Repo/rgc-imaging/other_packages/UKB_Brain_Pipeline/crash-20240325-101301-karl.landheer-mrt_romeo_combine-phase-ef840f93-8243-4e1c-b6b1-48b222c0e46c.pklz
Traceback (most recent call last):
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node mrt_romeo_combine-phase.

Cmdline:
	/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/scripts/romeo_unwrapping.jl --no-rescale --compute-B0 B0.nii --phase /Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mrt_romeo_combine-phase/multi-echo-phase.nii --mag /Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mrt_romeo_combine-phase/multi-echo-mag.nii -t '[9.42 19.7]' --phase-offset-correction monopolar
Stdout:

Stderr:
	env: julia: No such file or directory
Traceback:
	Traceback (most recent call last):
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nibabel/loadsave.py", line 100, in load
	    stat_result = os.stat(filename)
	FileNotFoundError: [Errno 2] No such file or directory: 'unwrapped.nii'

	During handling of the above exception, another exception occurred:

	Traceback (most recent call last):
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/interfaces/base/core.py", line 400, in run
	    outputs = self.aggregate_outputs(runtime)
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/interfaces/base/core.py", line 429, in aggregate_outputs
	    predicted_outputs = self._list_outputs()  # Predictions from _list_outputs
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/interfaces/nipype_interface_romeo.py", line 97, in _list_outputs
	    outputs['phase_unwrapped'] = split_multi_echo("unwrapped.nii", [extend_fname(f, "_romeo-unwrapped", ext="nii", out_dir=os.getcwd()) for f in self.inputs.phase])
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/interfaces/nipype_interface_romeo.py", line 20, in split_multi_echo
	    image4d = nib.load(in_path).get_fdata()
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nibabel/loadsave.py", line 102, in load
	    raise FileNotFoundError(f"No such file or no access: '{filename}'")
	FileNotFoundError: No such file or no access: 'unwrapped.nii'


240325-10:13:01,705 nipype.workflow ERROR:
	 Node romeo-voxelquality failed to run on host C02F203JMD6R.
240325-10:13:01,706 nipype.workflow ERROR:
	 Saving crash info to /Users/karl.landheer/PythonProjects/RGCi_Repo/rgc-imaging/other_packages/UKB_Brain_Pipeline/crash-20240325-101301-karl.landheer-romeo-voxelquality-db1a0fac-8489-4718-b395-6d4fd0b6e1ad.pklz
Traceback (most recent call last):
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node romeo-voxelquality.

Cmdline:
	/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/scripts/romeo_voxelquality.jl --TEs '[0.00942]' --mag /Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mask_workflow/nibabal-numpy_combine-magnitude/sub-1_ses-1_acq-swi_echo-1_part-mag_MEGRE_resampled_combined.nii.gz --phase /Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/nibabel_numpy_nilearn_axial-resampling/mapflow/_nibabel_numpy_nilearn_axial-resampling0/sub-1_ses-1_acq-swi_echo-1_part-phase_MEGRE_scaled_resampled.nii.gz --output sub-1_ses-1_acq-swi_echo-1_part-phase_MEGRE_scaled_resampled_romeo_voxelquality.nii --type grad+second+mag_weight+mag_coherence
Stdout:

Stderr:
	env: julia: No such file or directory
Traceback:
	RuntimeError: subprocess exited with code 127.

240325-10:13:01,709 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 0 jobs ready. Free memory (GB): 14.40/14.40, Free processors: 12/12.
240325-10:13:03,700 nipype.workflow INFO:
	 ***********************************
240325-10:13:03,701 nipype.workflow ERROR:
	 could not run node: workflow.sub-1.ses-1.qsmxt_acq-swi.mrt_romeo_combine-phase
240325-10:13:03,701 nipype.workflow INFO:
	 crashfile: /Users/karl.landheer/PythonProjects/RGCi_Repo/rgc-imaging/other_packages/UKB_Brain_Pipeline/crash-20240325-101301-karl.landheer-mrt_romeo_combine-phase-ef840f93-8243-4e1c-b6b1-48b222c0e46c.pklz
240325-10:13:03,702 nipype.workflow ERROR:
	 could not run node: workflow.sub-1.ses-1.qsmxt_acq-swi.mask_workflow.romeo-voxelquality
240325-10:13:03,703 nipype.workflow INFO:
	 crashfile: /Users/karl.landheer/PythonProjects/RGCi_Repo/rgc-imaging/other_packages/UKB_Brain_Pipeline/crash-20240325-101301-karl.landheer-romeo-voxelquality-db1a0fac-8489-4718-b395-6d4fd0b6e1ad.pklz
240325-10:13:03,704 nipype.workflow INFO:
	 ***********************************
RuntimeError: Traceback (most recent call last):
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node mrt_romeo_combine-phase.

Cmdline:
	/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/scripts/romeo_unwrapping.jl --no-rescale --compute-B0 B0.nii --phase /Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mrt_romeo_combine-phase/multi-echo-phase.nii --mag /Users/karl.landheer/Downloads/qsm_output/workflow/sub-1/ses-1/qsmxt_acq-swi/mrt_romeo_combine-phase/multi-echo-mag.nii -t '[9.42 19.7]' --phase-offset-correction monopolar
Stdout:

Stderr:
	env: julia: No such file or directory
Traceback:
	Traceback (most recent call last):
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nibabel/loadsave.py", line 100, in load
	    stat_result = os.stat(filename)
	FileNotFoundError: [Errno 2] No such file or directory: 'unwrapped.nii'

	During handling of the above exception, another exception occurred:

	Traceback (most recent call last):
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/interfaces/base/core.py", line 400, in run
	    outputs = self.aggregate_outputs(runtime)
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/interfaces/base/core.py", line 429, in aggregate_outputs
	    predicted_outputs = self._list_outputs()  # Predictions from _list_outputs
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/interfaces/nipype_interface_romeo.py", line 97, in _list_outputs
	    outputs['phase_unwrapped'] = split_multi_echo("unwrapped.nii", [extend_fname(f, "_romeo-unwrapped", ext="nii", out_dir=os.getcwd()) for f in self.inputs.phase])
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/interfaces/nipype_interface_romeo.py", line 20, in split_multi_echo
	    image4d = nib.load(in_path).get_fdata()
	  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nibabel/loadsave.py", line 102, in load
	    raise FileNotFoundError(f"No such file or no access: '{filename}'")
	FileNotFoundError: No such file or no access: 'unwrapped.nii'



The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/karl.landheer/opt/anaconda3/envs/UKB/bin/qsmxt", line 8, in <module>
    sys.exit(main())
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/qsmxt/cli/main.py", line 1561, in main
    wf.run(
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/engine/workflows.py", line 638, in run
    runner.run(execgraph, updatehash=updatehash, config=self.config)
  File "/Users/karl.landheer/opt/anaconda3/envs/ukb/lib/python3.8/site-packages/nipype/pipeline/plugins/base.py", line 224, in run
    raise error from cause
RuntimeError: 2 raised. Re-raising first.

**Comment by karllandheer** [comment]
I think I am just missing Julia...

**Comment by karllandheer** [comment]
Hello, I had issues with it even after I installed Julia 1.9, but anyways it worked fine from the docker, I think. This is an example output, does this look right?

![image](https://github.com/QSMxT/QSMxT/assets/76697225/fbc9fe81-afcf-4816-8745-08fb7311e725)

It looks like there may have been an issue with the add_bet part? I have the brain masks from a separate part of the pipeline, is there a way to add it in without relying on the bet tool? Would manually masking the mag/phase images by the brain mask work, or is it not that simple?

As an aside, will you be in Singapore? I would be interested to learn more about qsmxt

**Comment by karllandheer** [comment]
I'm not so sure this is right. One thing is that I am working from the raw dicom data, not the data you suggested (the reason being those are already processed and many are missing). The raw data is available here: https://biobank.ndph.ox.ac.uk/showcase/refer.cgi?id=32579

If you unpack it it produces an image (magnitude and phase) for each of the 32 coils. This data is then processed into the PHASE_TE1 / PHASE_TE2, etc, that you used by a pipeline. This pipeline does things like coil combination, phase unwrapping via fsl, etc. I could implement this pipeline, but I am wondering if qsmxt is able to handle multi coil data, or if you have such a tool available? Or is it expected that the images should be in this coil combined and phase-unwrapped form?

**Comment by astewartau** [comment]
## `[Errno 13] Permission denied: 'Users/karl.json`
Thanks very much for flagging this! It is a bug in QSMxT, which I have just resolved in the main branch. The fix will be included in v6.4.3, which will be released soon. I'm glad you found a workaround for now. 

## `env: julia: No such file or directory`
As you have discovered, it looks like you were missing Julia. Indeed, I recommend using the containerised version of QSMxT to avoid these dependency problems. In case you are interested, there is a detailed list of requirements in the [container build script](https://github.com/NeuroDesk/neurocontainers/blob/master/recipes/qsmxt/build.sh), including a series of Julia packages described [alongside it](https://github.com/NeuroDesk/neurocontainers/blob/master/recipes/qsmxt/install_packages.jl), which may explain whatever the additional problem you encountered was (possibly you were missing these Julia packages). 

## Output
Your output indeed does not look perfect and has some intrusive artefacts, unlike the example that I processed earlier. What was the exact command that you used to generate it? Could you provide some of the output logs to confirm the version etc.? If it is exactly the same as how I processed the other example, this dataset may be an outlier or have different characteristics. I would consider trying `--premade bet` to see if the result is any better - this pipeline uses a more straightforward masking algorithm. 

In case it provides any further insight; this is what my input magnitude and phase images look like from my BIDS directory:

![image](https://github.com/QSMxT/QSMxT/assets/9957135/bbabc12d-b7de-4bf1-a359-ef561e991fd1)

## Coil combination
Unfortunately, while the DICOM images include combined and uncombined magnitude and phase data, the phase data is incorrectly combined. So you are right that you would need to run a coil combination algorithm separately on the uncombined images since QSMxT expects already-combined images.

Including coil combination in QSMxT is something I have considered, but there is also no standard way to store uncombined coil data in BIDS as inputs, although an [extension](https://docs.google.com/document/d/1kyw9mGgacNqeMbp4xZet3RnDhcMmf4_BmRgKaOkO2Sc/edit) to support it was proposed which we could use. 

If you want to run a separate coil combination, you could also consider ASPIRE/MCPC-3D-S, which also conveniently runs in Julia (see [korbinian90/MriResearchTools.jl](https://github.com/korbinian90/MriResearchTools.jl)). You could then convert the output to BIDS to run in QSMxT. This is what I am considering integrating into QSMxT at some stage. 

About the NIfTI data I suggested - when you say that many are missing, did you look into both the [26301](https://biobank.ndph.ox.ac.uk/showcase/field.cgi?tk=ZAMoKBtMR9UCLJwFLdoQTGoOR1WdfDKG406504&id=26301) and [20251](https://biobank.ndph.ox.ac.uk/showcase/field.cgi?tk=ZAMoKBtMR9UCLJwFLdoQTGoOR1WdfDKG406504&id=20251) data fields (since both are required)? Or do many subjects not have this?

**Comment by astewartau** [comment]
Oh, and I forgot to mention that I will be in Singapore! I am presenting work about evaluating QSM and ideas for a future 'always-online' QSM challenge (prototype [here](https://github.com/qsmxt/qsm-ci)). It would be great to meet you. :smile:

**Comment by karllandheer** [comment]
Hello again, I think I solved and found the issue. I was using the phase maps from the scanner, but I reconstructed them using the MCPC-3D-S coil combination method I found in the literature. This is an example of the resulting QSM maps now:

<img width="334" alt="Screenshot 2024-03-26 at 2 28 58 PM" src="https://github.com/QSMxT/QSMxT/assets/76697225/ba4e557f-542c-45d4-b202-5fb63c102d96">

Does this look right? I think so but not sure. A few remaining quick questions:

1) The scale seems to be substantially different from the values from the UKB pipeline. Perhaps this is just a difference in the units. For example the maps I have from qsmxt seem have a value in the csf of ~0.015. The UKB QSM pipeline has a value of ~90. Is the values on my end correct or am I still incorrectly specifying something?

2) Do you have a poster or talk at ISMRM? If so please let me know the time and I will do my best to make it.

**Comment by astewartau** [comment]
That updated reconstruction looks a lot better! This subject appears to have larger ventricles - the larger amount of free fluid may lead to noisier phase values which different algorithms may respond to differently depending on settings such as regularisation. Overall, this reconstruction looks much more reasonable. 

1. The scale of QSMxT's images is parts-per-million (ppm), whereas the BioBank data uses parts-per-billion (ppb). You can see how they scale it [here](https://git.fmrib.ox.ac.uk/cwang/uk_biobank_qsm_pipeline/-/blob/main/QSM_only_scripts_MATLAB/UKBiobank_QSM.m?ref_type=heads#L225). In the same units, QSMxT is producing 15 vs. UKB's 90 for CSF, so there is still some difference. By default, QSMxT also references QSM values with respect to the mean susceptibility value across the brain. Interestingly, UKB is supposed to have referenced QSM values to CSF, so we would expect the CSF value to be closer to zero in theirs... If you used QSMxT's pipeline for segmentation and analysis, which is produced using FastSurfer for segmentation and ANTs for registration, this may also have an effect since UKB used FSL's FAST for segmentation and FIRST for registration. That is not to say that any particular result is more correct, but that this could explain some of the differences in addition to the trade-offs that different QSM algorithms make. 

2. I have a poster presentation at ISMRM titled "QSM-CI: An automated continuous QSM challenge" (abstract #6388). It will be during the 'Quantitative Susceptibility Mapping' poster session on the 6th of May at 9:15 AM. I also plan to attend the EMTP study group session, which I may have the opportunity to present at, depending on available spots.

**Comment by astewartau** [comment]
Wonderful! 😊 I will close this issue for now and reopen if you have further concerns on this topic.
